lib/mllp/client.ex

```diff
@@ -0,0 +1,444 @@
+defmodule MLLP.ClientContract do
+  @type error_type :: :connect_failure | :send_error | :recv_error
+  @type error_reason :: :closed | :timeout | :no_socket | :inet.posix()
+
+  @type sender_error :: MLLP.Client.Error.t()
+
+  @type options :: [
+          expect_reply: boolean(),
+          timeout: non_neg_integer(),
+          wrap_message: boolean(),
+          reply_timeout: non_neg_integer() | :infinity,
+          verify_reply: boolean(),
+          tls_opts: map(),
+          socket_opts: map()
+        ]
+
+  @type send_options :: %{
+          optional(:expect_reply) => boolean(),
+          optional(:timeout) => non_neg_integer(),
+          optional(:wrap_message) => boolean(),
+          optional(:reply_timeout) => non_neg_integer() | :infinity,
+          optional(:verify_reply) => boolean()
+        }
+
+  @callback send(
```

Comment:
Couple of points on this:

- Should we have two methods `send_async` and `send`
     - send_async - as the name suggest does not wait for reply but can send reply using a callback. It also takes away `exact_reply` parameter
     - send - is synchronous and waiting for reply

On `verify_reply` - It's about checking reply if its valid HL7 ACK or not. Should we do it only if payload is `HL7.Message.t()`, otherwise we don't need to do it. And we can remove this flag also.

On `wrap_message` - Should we do it if message is not wrapped without using the flag?

Reply:
> Should we have two methods `send_async` and `send`
> 
>     * send_async - as the name suggest does not wait for reply but can send reply using a callback. It also takes away `exact_reply` parameter
> 
>     * send - is synchronous and waiting for reply


I  like  that idea a lot, I very much don't like all the options that can be manipulated. The less options the better, but see below. 

> On verify_reply - It's about checking reply if its valid HL7 ACK or not. Should we do it only if payload is HL7.Message.t(), otherwise we don't need to do it. And we can remove this flag also.

That makes sense  to me. I simply didn't know all the cases we would want to support. In that regard, I think `send_async` is not even really needed per the MLLP protocol itself. Can  you think of an actual case where a caller would want to sent and not receive a reply? I  can not. 

> On wrap_message - Should we do it if message is not wrapped without using the flag?

I thought about that, that makes perfect sense to me,  but it comes back to the same question.  Are  there actually any cases where someone would send data that's not enveloped? I think a lof  of the decisions I made around options was merely to support the existing test cases, but I didn't feel great about them.

Reply:
`wrap_message` -  Most caller shouldn't worry about wrapping message at their end. It's a MLLP protocol thing and I think it should be handled here.
`send_async` - I haven't seen this case either, but since we have its implementation, we can keep it. It may be useful in cases where someone wants to just blast off messages without worrying about ACKs during some sort of test.

Reply:
Should  we always  unwrap per getting rid of wrap_message option? Makes sense to me.